File: backend/app/schemas/neighbourhood.py
from pydantic import (
    BaseModel,
    Field,
    StringConstraints,
    model_validator,
)
from typing import Annotated, Literal
from uuid import UUID
from enum import Enum
from datetime import datetime

from app.models.security_officer import AvailabilityStatus
from app.models.neighbourhood_user import NeighbourhoodRole
from math import isfinite

NonEmptyString = Annotated[str, StringConstraints(min_length=1, strip_whitespace=True)]

class CreateNeighbourhoodReq(BaseModel):
    name: NonEmptyString
    location: NonEmptyString
    property_id: UUID

class NeighbourhoodRes(BaseModel):
    id: UUID
    name: NonEmptyString
    location: NonEmptyString   
    join_code: NonEmptyString
    created_at: datetime

class CreateNeighbourhoodRes(BaseModel):     
    status: int
    message: str | None = None
    data: NeighbourhoodRes | None = None

class NeighbourhoodPropertyRes(BaseModel):
    id: UUID
    address: NonEmptyString
    property_type: Literal["PRIVATE", "PUBLIC"]
    neighbourhood_id: UUID | None = None
    neighbourhood_name: str | None = None

class NeighbourhoodMapPropertyRes(BaseModel):
    id: UUID
    address: NonEmptyString
    property_type: Literal["PRIVATE", "PUBLIC"]
    latitude: float | None = Field(default=None, ge=-90, le=90)
    longitude: float | None = Field(default=None, ge=-180, le=180)


class UpdateMemberRoleReq(BaseModel):
    role: NeighbourhoodRole


class NeighbourhoodMemberRes(BaseModel):
    user_id: UUID
    first_name: str
    last_name: str
    email: str
    role: NeighbourhoodRole


class UpdateMemberRoleRes(BaseModel):
    status: int
    message: str
    data: NeighbourhoodMemberRes


class GetSecurityAvailabilityRes(BaseModel):
    status: int
    message: str | None = None
    availability: AvailabilityStatus | None
    location_updated_at: datetime | None

class OnDutyStatus(str, Enum):
    ON_DUTY = "ON_DUTY"
    OFF_DUTY = "OFF_DUTY"

class UpdateSecurityAvailabilityReq(BaseModel):
    neighbourhood_id: UUID
    new_duty_status: OnDutyStatus

class UpdateSecurityAvailabilityRes(BaseModel):
    status: int
    message: str | None = None

class UpdateOfficerLocationReq(BaseModel):
    neighbourhood_id: UUID
    latitude: float = Field(..., ge=-90, le=90)
    longitude: float = Field(..., ge=-180, le=180)

class UpdateOfficerLocationRes(BaseModel):
    status: int
    message: str | None = None

class OfficerLocationRes(BaseModel):
    officer_id: UUID
    latitude: float
    longitude: float
    location_updated_at: datetime
    is_stale: bool


class NeighbourhoodBoundaryPolygon(BaseModel):
    type: Literal["Polygon"]
    coordinates: list[list[tuple[float, float]]] = Field(
        min_length=1,
    )

    @model_validator(mode="after")
    def validate_polygon(self):
        for ring in self.coordinates:
            if len(ring) < 4:
                raise ValueError(
                    "Each polygon ring must contain at least four points"
                )

            if ring[0] != ring[-1]:
                raise ValueError(
                    "The polygon ring must be closed"
                )

            for longitude, latitude in ring:
                if not isfinite(longitude):
                    raise ValueError(
                        "Longitude must be finite"
                    )

                if not isfinite(latitude):
                    raise ValueError(
                        "Latitude must be finite"
                    )

                if not -180 <= longitude <= 180:
                    raise ValueError(
                        "Longitude must be between -180 and 180"
                    )

                if not -90 <= latitude <= 90:
                    raise ValueError(
                        "Latitude must be between -90 and 90"
                    )

        return self


class NeighbourhoodBoundaryImportReq(
    NeighbourhoodBoundaryPolygon,
):
    pass


class NeighbourhoodBoundaryData(BaseModel):
    neighbourhood_id: UUID
    geometry: NeighbourhoodBoundaryPolygon


class NeighbourhoodBoundaryRes(BaseModel):
    status: int
    message: str | None = None
    data: NeighbourhoodBoundaryData | None = None

File: frontend/src/lib/validators/neighbourhood.ts
import { z } from "zod";

export const CreateNeighbourhoodReqSchema = z.object({
  name: z.string().min(1, "Name is required"),
  location: z.string().min(1, "Location is required"),
  property_id: z.string().regex(/^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$/i, "Invalid UUID format"),
});

export const NeighbourhoodResSchema = z.object({
  id: z.string().regex(/^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$/i, "Invalid UUID format"),
  name: z.string().min(1),
  location: z.string().min(1),
  join_code: z.string().min(1),
  created_at: z.coerce.date(),
});

export const CreateNeighbourhoodResSchema = z.object({
  status: z.number().int(),
  message: z.string().nullable().optional(),
  data: NeighbourhoodResSchema.nullable().optional(),
});

export const NeighbourPropertychema = z.object({
  id: z.string(),
  address: z.string(),
  property_type: z.enum(["PUBLIC", "PRIVATE"]),
  neighbourhood_id: z.string().nullable(),
  neighbourhood_name: z.string().optional(),
})

export const NeighbourPropertiesResSchema = z.array(NeighbourPropertychema)

export const NeighbourhoodMemberResSchema = z.object({
  user_id: z.string().uuid(),
  first_name: z.string(),
  last_name: z.string(),
  email: z.string().email(),
  role: z.enum([
    "RESIDENT",
    "SECURITY_OFFICER",
    "NEIGHBOURHOOD_ADMIN"
  ])
});

export const NeighbourhoodMembersResSchema = z.array(NeighbourhoodMemberResSchema);

export const UpdateMemberRoleReqSchema = z.object({
  role: z.enum([
    "RESIDENT",
    "SECURITY_OFFICER",
    "NEIGHBOURHOOD_ADMIN"
  ])
});

export const UpdateMemberRoleResSchema = z.object({
  status: z.number().int(),
  message: z.string(),
  data: NeighbourhoodMemberResSchema
});

export const LeaveNeighbourhoodParamsSchema = z.object({
  neighbourhoodId: z.uuid("Invalid neighbourhood ID"),
  propertyId: z.uuid("Invalid property ID")
});

export const AvailabilityStatusSchema = z.enum(["AVAILABLE", "BUSY", "UNAVAILABLE"]);

export const UpdateSecurityAvailabilityReqSchema = z.object({
  neighbourhood_id: z.uuid("Invalid neighbourhood ID"),
  new_availability: AvailabilityStatusSchema,
});

export const UpdateSecurityAvailabilityResSchema = z.object({
  status: z.number().int(),
  message: z.string().nullable().optional(),
});

export const GetSecurityAvailabilityResSchema = z.object({
  status: z.number().int(),
  message: z.string().nullable().optional(),
  availability: AvailabilityStatusSchema,
  location_updated_at: z.coerce.date(),
});

export const UpdateSecurityLocationReqSchema = z.object({
  neighbourhood_id: z.uuid("Invalid neighbourhood ID"),
  latitude: z.number().min(-90).max(90),
  longitude: z.number().min(-180).max(180),
});

export const UpdateSecurityLocationResSchema = z.object({
  status: z.number().int(),
  message: z.string().nullable().optional(),
});

export const OfficerLocationResSchema = z.object({
  officer_id: z.uuid("Invalid officer ID"),
  latitude: z.number().min(-90).max(90),
  longitude: z.number().min(-180).max(180),
  location_updated_at: z.coerce.date(),
  is_stale: z.boolean(),
});

export type CreateNeighbourhoodReq = z.infer<typeof CreateNeighbourhoodReqSchema>;
export type NeighbourhoodRes = z.infer<typeof NeighbourhoodResSchema>;
export type CreateNeighbourhoodRes = z.infer<typeof CreateNeighbourhoodResSchema>;
export type NeighbourPropertiesRes = z.infer<typeof NeighbourPropertiesResSchema>;
export type NeighbourhoodMemberRes = z.infer<typeof NeighbourhoodMemberResSchema>;
export type NeighbourhoodMembersRes = z.infer<typeof NeighbourhoodMembersResSchema>;
export type UpdateMemberRoleReq = z.infer<typeof UpdateMemberRoleReqSchema>;
export type UpdateMemberRoleRes = z.infer<typeof UpdateMemberRoleResSchema>;
export type LeaveNeighbourhoodParams = z.infer<typeof LeaveNeighbourhoodParamsSchema>;
export type AvailabilityStatus = z.infer<typeof AvailabilityStatusSchema>;
export type UpdateSecurityAvailabilityReq = z.infer<typeof UpdateSecurityAvailabilityReqSchema>;
export type UpdateSecurityAvailabilityRes = z.infer<typeof UpdateSecurityAvailabilityResSchema>;
export type GetSecurityAvailabilityRes = z.infer<typeof GetSecurityAvailabilityResSchema>;
export type UpdateSecurityLocationRes = z.infer<typeof UpdateSecurityLocationResSchema>;
export type UpdateSecurityLocationReq = z.infer<typeof UpdateSecurityLocationReqSchema>;
export type OfficerLocationRes = z.infer<typeof OfficerLocationResSchema>;
